tower-balance/src/lib.rs

```diff
@@ -26,166 +83,185 @@ pub enum Error<T, U> {
+    {
+        match self.ready.get_index_mut(idx) {
+            None => return None,
+            Some((_, svc)) => {
+                match svc.poll_ready() {
+                    Ok(Async::Ready(())) => return Some(Ok(Async::Ready(()))),
+                    Err(e) => return Some(Err(Error::Inner(e))),
+                    Ok(Async::NotReady) => {}
+                }
+            }
+        }
 
-    fn update_endpoints(&mut self) -> Result<(), T::DiscoverError> {
-        use tower_discover::Change::*;
+        let (key, svc) = self.ready.swap_remove_index(idx).expect("invalid ready index");
+        self.not_ready.insert(key, svc);
+        Some(Ok(Async::NotReady))
+    }
 
+    /// Chooses the next service to which a request will be dispatched.
+    ///
+    /// Ensures that .
+    fn choose_and_poll_ready(&mut self)
+        -> Poll<(), Error<<D::Service as Service>::Error, D::DiscoverError>>
+    {
         loop {
-            let change = match try!(self.discover.poll()) {
-                Async::Ready(change) => change,
-                Async::NotReady => return Ok(()),
+            let idx = match self.ready.len() {
+                0 => return Ok(Async::NotReady),
+                1 => 0,
+                _ => self.choose.call(&self.ready),
             };
 
-            match change {
-                Insert(key, val) => {
-                    // TODO: What to do if there already is an entry with the
-                    // given `key` in the set?
-                    self.endpoints.entry(key).or_insert(val);
-                }
-                Remove(key) => {
-                    // TODO: What to do if there is no entry?
-                    self.endpoints.remove(&key);
-                }
+            // XXX Should we handle per-endpoint errors?
+            if self.poll_ready_index(idx).expect("invalid ready index")?.is_ready() {
+                self.chosen_ready_index = Some(idx);
+                return Ok(Async::Ready(()));
             }
+        }
+    }
+}
 
-            // For now, we're just going to reset the pos to zero, but this is
-            // definitely not ideal.
-            //
-            // TODO: improve
-            self.pos = 0;
+impl<D, C> Service for Balance<D, C>
+where
+    D: Discover,
+    C: Choose,
+    D::Service: Load<Metric = C::Metric>,
+{
+    type Request = <D::Service as Service>::Request;
+    type Response = <D::Service as Service>::Response;
+    type Error = Error<<D::Service as Service>::Error, D::DiscoverError>;
+    type Future = ResponseFuture<<D::Service as Service>::Future, D::DiscoverError>;
+
+    /// Prepares the balancer to process a request.
+    ///
+    /// When `Async::Ready` is returned, `chosen_ready_index` is set with a valid index
+    /// into `ready` referring to a `Service` that is ready to disptach a request.
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        // Clear before `ready` is altered.
+        self.chosen_ready_index = None;
```

Comment:
This is intended. If it's called again, we might update `ready` below, and this index would no longer be reliable (because the ordermap may change its order).